### ISSUE TITLE
[solved] how to train fine-tuning classification model (size mismatch for head.bias: copying a param with shape torch.Size([1000]) from checkpoint, the shape in current model is torch.Size([4]).)

### ISSUE THREAD

**Issue by jeonga0303** [OpenGVLab/InternImage#297]
<img width="491" alt="image" src="https://github.com/OpenGVLab/InternImage/assets/52391624/42544e5a-696b-42ab-8a2a-a2a1c0f22441">

I customized the config.py.
how to train fine-tuning classification model?

<img width="897" alt="image" src="https://github.com/OpenGVLab/InternImage/assets/52391624/a0f1db42-90d4-4927-b828-f3b1905fb240">

**Comment by jeonga0303** [comment]
If you want to learn the image classification model using the pretrained model, proceed to the next step

utils.py If you add the following code block under line 240, it's possible
`        if (Nc1 != Nc2):
            model.head.weight.data.normal_(mean=0.0, std=0.01)
            model.head.bias.data.zero_()
            del state_dict['head.weight']
            del state_dict['head.bias']
            logger.warning(
                f'Error in loading classifier head, re-init classifier head to 0'
            )
`

And config.py updates the num_classes information to the number of classes in your dataset
<img width="407" alt="image" src="https://github.com/OpenGVLab/InternImage/assets/52391624/01198f01-190c-47f1-8217-1bb9b147a970">
Of course, the pre-trained model needs to be downloaded in advance

**Comment by jeonga0303** [comment]
If you want to learn the image classification model using the pretrained model, proceed to the next step

utils.py If you add the following code block under line 240, it's possible
`        if (Nc1 != Nc2):
            model.head.weight.data.normal_(mean=0.0, std=0.01)
            model.head.bias.data.zero_()
            del state_dict['head.weight']
            del state_dict['head.bias']
            logger.warning(
                f'Error in loading classifier head, re-init classifier head to 0'
            )
`

And config.py updates the num_classes information to the number of classes in your dataset
<img width="407" alt="image" src="https://github.com/OpenGVLab/InternImage/assets/52391624/01198f01-190c-47f1-8217-1bb9b147a970">
Of course, the pre-trained model needs to be downloaded in advance

**Comment by jeonga0303** [comment]
how to convert Nc1..

<img width="573" alt="image" src="https://github.com/OpenGVLab/InternImage/assets/52391624/84529ae9-dd85-47c4-9d03-d13afcb0b702">
<img width="142" alt="image" src="https://github.com/OpenGVLab/InternImage/assets/52391624/cc583ccd-3c70-48f7-9703-e8d1c7f6724f">

<img width="898" alt="image" src="https://github.com/OpenGVLab/InternImage/assets/52391624/fb4d405b-027f-4e74-8050-e3138c405f27">

**Comment by jeonga0303** [comment]
I tried changing the file configuration in the following order.

I'm training, but the data is big, so I'll let you know the results in the future

1. download pth file

2. config.py

```
_C.DATA.IMG_SIZE = 640
_C.MODEL.PRETRAINED = 'internimage_h_22kto1k_640.pth'
_C.MODEL.NUM_CLASSES = 4

```

3. util.py
convert load_pretrained function.

```
def load_pretrained(config, model, logger):
    logger.info(f'==============> Loading weight {config.MODEL.PRETRAINED} for fine-tuning......')
    checkpoint = torch.load(config.MODEL.PRETRAINED, map_location='cpu')

    state_dict = checkpoint
    if 'model' in checkpoint:
        state_dict = checkpoint['model']
    elif 'module' in checkpoint:
        state_dict = checkpoint['module']

    first_key = list(state_dict.keys())[0]
    # delete teacher weights
    if 'student' in first_key or 'teacher' in first_key:
        new_state_dict = OrderedDict()
        for k, v in state_dict.items():
            if 'student_proj' in k:
                continue
            if 'student' in k:
                new_k = k.replace('student.', '')
                new_state_dict[new_k] = v
        state_dict = new_state_dict

    # weights from sim
    if 'mask_token' in first_key:
        new_state_dict = OrderedDict()
        for k, v in state_dict.items():
            if 'mm_dcnv3' in k:
                continue
            if 'dcnv3' not in k and 'clip_projector' not in k:
                continue
            new_k = k.replace('dcnv3.', '')
            new_state_dict[new_k] = v
        new_state_dict['fc_norm.weight'] = state_dict['clip.classifier_ln.weight']
        new_state_dict['fc_norm.bias'] = state_dict['clip.classifier_ln.bias']
        new_state_dict['head.weight'] = state_dict['clip.classifier.weight']
        new_state_dict['head.bias'] = state_dict['clip.classifier.bias']
        state_dict = new_state_dict

    # delete relative_position_index since we always re-init it
    relative_position_index_keys = [k for k in state_dict.keys() if 'relative_position_index' in k]
    for k in relative_position_index_keys:
        del state_dict[k]

    # delete relative_coords_table since we always re-init it
    relative_position_index_keys = [k for k in state_dict.keys() if 'relative_coords_table' in k]
    for k in relative_position_index_keys:
        del state_dict[k]

    # delete attn_mask since we always re-init it
    attn_mask_keys = [k for k in state_dict.keys() if 'attn_mask' in k]
    for k in attn_mask_keys:
        del state_dict[k]

    # bicubic interpolate relative_position_bias_table if not match
    relative_position_bias_table_keys = [k for k in state_dict.keys() if 'relative_position_bias_table' in k]
    for k in relative_position_bias_table_keys:
        relative_position_bias_table_pretrained = state_dict[k]
        relative_position_bias_table_current = model.state_dict()[k]
        L1, nH1 = relative_position_bias_table_pretrained.size()
        L2, nH2 = relative_position_bias_table_current.size()
        if nH1 != nH2:
            logger.warning(f'Error in loading {k}, passing......')
        else:
            if L1 != L2:
                # bicubic interpolate relative_position_bias_table if not match
                S1 = int(L1**0.5)
                S2 = int(L2**0.5)
                relative_position_bias_table_pretrained_resized = torch.nn.functional.interpolate(
                    relative_position_bias_table_pretrained.permute(1, 0).view(1, nH1, S1, S1),
                    size=(S2, S2),
                    mode='bicubic')
                state_dict[k] = relative_position_bias_table_pretrained_resized.view(nH2, L2).permute(1, 0)

    # bicubic interpolate absolute_pos_embed if not match
    absolute_pos_embed_keys = [k for k in state_dict.keys() if 'absolute_pos_embed' in k]
    for k in absolute_pos_embed_keys:
        # dpe
        absolute_pos_embed_pretrained = state_dict[k]
        absolute_pos_embed_current = model.state_dict()[k]
        _, L1, C1 = absolute_pos_embed_pretrained.size()
        _, L2, C2 = absolute_pos_embed_current.size()
        if C1 != C2:
            logger.warning(f'Error in loading {k}, passing......')
        else:
            if L1 != L2:
                S1 = int(L1**0.5)
                S2 = int(L2**0.5)
                absolute_pos_embed_pretrained = absolute_pos_embed_pretrained.reshape(-1, S1, S1, C1)
                absolute_pos_embed_pretrained = absolute_pos_embed_pretrained.permute(0, 3, 1, 2)
                absolute_pos_embed_pretrained_resized = torch.nn.functional.interpolate(
                    absolute_pos_embed_pretrained,
                    size=(S2, S2),
                    mode='bicubic')
                absolute_pos_embed_pretrained_resized = absolute_pos_embed_pretrained_resized.permute(0, 2, 3, 1)
                absolute_pos_embed_pretrained_resized = absolute_pos_embed_pretrained_resized.flatten(1, 2)
                state_dict[k] = absolute_pos_embed_pretrained_resized

    # check classifier, if not match, then re-init classifier to zero
    if 'head.bias' in state_dict:
        head_bias_pretrained = state_dict['head.bias']
        Nc1 = head_bias_pretrained.shape[0]
        Nc2 = model.head.bias.shape[0]
        logger.info(f'Nc1, Nc2: {Nc1}, {Nc2}')
        if Nc1 != Nc2:
            logger.warning(f'Error in loading classifier head, re-initializing classifier head to match dimensions.')
            model.head.weight = torch.nn.Parameter(torch.empty(Nc2, model.head.weight.shape[1]).uniform_(-0.1, 0.1))
            model.head.bias = torch.nn.Parameter(torch.empty(Nc2).uniform_(-0.1, 0.1))
            del state_dict['head.weight']
            del state_dict['head.bias']

    if config.MODEL.INTERN_IMAGE.REMOVE_CENTER:
        for k, v in state_dict.items():
            if 'dcn.mask.bias' in k:
                idx = list(range(v.shape[0]))
                idx = [i for i in idx if i != 4 and (i - 4) % 9 != 0]
                state_dict[k] = v[idx]
            if 'dcn.offset.bias' in k:
                idx = list(range(v.shape[0]))
                idx = [i for i in idx if i != 8 and (i - 8) % 18 != 0 and i != 9 and (i - 9) % 18 != 0]
                state_dict[k] = v[idx]
            if 'dcn.mask.weight' in k:
                idx = list(range(v.shape[0]))
                idx = [i for i in idx if i != 4 and (i - 4) % 9 != 0]
                state_dict[k] = v[idx, :]
            if 'dcn.offset.weight' in k:
                idx = list(range(v.shape[0]))
                idx = [i for i in idx if i != 8 and (i - 8) % 18 != 0 and i != 9 and (i - 9) % 18 != 0]
                state_dict[k] = v[idx, :]

    msg = model.load_state_dict(state_dict, strict=False)
    logger.warning(msg)

    logger.info(f'=> loaded successfully {config.MODEL.PRETRAINED}')

    del checkpoint
    torch.cuda.empty_cache()
```

4. dataset/samplers.py
convert iteration.

```
def __iter__(self):
        # deterministically shuffle based on epoch
        g = torch.Generator()
        g.manual_seed(self.epoch)

        t = torch.Generator()
        t.manual_seed(0)

        indices = torch.randperm(len(self.dataset), generator=t).tolist()
        indices = [i for i in indices if i % self.num_parts == self.rank]

        # add extra samples to make it evenly divisible
        while len(indices) < self.total_size_parts:
            indices += indices[:(self.total_size_parts - len(indices))]
        
        indices = indices[:self.total_size_parts]
        assert len(indices) == self.total_size_parts, f'Length of indices ({len(indices)}) does not match total_size_parts ({self.total_size_parts})'

        # subsample
        indices = indices[self.rank // self.num_parts:self.total_size_parts:self.num_replicas // self.num_parts]

        index = torch.randperm(len(indices), generator=g).tolist()
        indices = list(np.array(indices)[index])

        assert len(indices) == self.num_samples, f'Length of indices ({len(indices)}) does not match num_samples ({self.num_samples})'

        return iter(indices)
```

5. cmd
`python -m torch.distributed.launch --nproc_per_node [your gpu num] --master_port 12345 main_deepspeed.py --cfg  configs/without_lr_decay/internimage_h_22kto1k_640_custom.yaml  --batch-size 16 --accumulation-steps 4 --pretrained internimage_h_22kto1k_640.pth --data-path [dataset_path] --output work_dirs_deepspeed --local-rank 1`

my gpu is a100 * 2.

**Comment by jeonga0303** [comment]
[bug]

I don't think there's progress in training..
Everything's the same as loss
May I know the reason?

<img width="707" alt="image" src="https://github.com/OpenGVLab/InternImage/assets/52391624/7930cc93-74b0-4197-9395-7236c4463c0b">

**Comment by LovelySimon** [comment]
hello，how did you put the model on gpus? I got Segmentation fault (core dumped) when run model.cuda() on 3090gpu.So sad....
